ssl/statem/extensions_srvr.c

```diff
@@ -1946,3 +1946,153 @@ EXT_RETURN tls_construct_stoc_psk(SSL_CONNECTION *s, WPACKET *pkt,
+        sc->ext.client_cert_type_ctos = 0;
+        return EXT_RETURN_NOT_SENT;
+    }
+    /*
+     * Note: only supposed to send this if we are going to do a cert request,
+     * but TLSv1.3 could do a PHA request if the client supports it
+     */
+    if ((send_certificate_request(sc) || sc->post_handshake_auth == SSL_PHA_EXT_RECEIVED)
+            && sc->ext.client_cert_type_ctos
+            && sc->client_cert_type != NULL) {
+        if (!WPACKET_put_bytes_u16(pkt, TLSEXT_TYPE_client_cert_type)
+                || !WPACKET_start_sub_packet_u16(pkt)
+                || !WPACKET_put_bytes_u8(pkt, sc->ext.client_cert_type)
+                || !WPACKET_close(pkt)) {
+            SSLfatal(sc, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+            return EXT_RETURN_FAIL;
+        }
+        return EXT_RETURN_SENT;
+    }
+    /* if we don't send it, reset to TLSEXT_cert_type_x509 */
+    sc->ext.client_cert_type_ctos = 0;
+    sc->ext.client_cert_type = TLSEXT_cert_type_x509;
+    return EXT_RETURN_NOT_SENT;
+}
+
+/* One of |pref|, |other| is configured and the values are sanitized */
+static int reconcile_cert_type(const unsigned char *pref, size_t pref_len,
+                               const unsigned char *other, size_t other_len,
+                               uint8_t *chosen_cert_type)
+{
+    size_t i;
+
+    for (i = 0; i < pref_len; i++) {
+        if (memchr(other, pref[i], other_len) != NULL) {
+            *chosen_cert_type = pref[i];
+            return 1;
+        }
+    }
+    return 0;
+}
+
+int tls_parse_ctos_client_cert_type(SSL_CONNECTION *sc, PACKET *pkt,
+                                    unsigned int context,
+                                    X509 *x, size_t chainidx)
+{
+    PACKET supported_cert_types;
+    const unsigned char *data;
+    size_t len;
+
+    /* Ignore the extension */
+    if (sc->client_cert_type == NULL) {
+        sc->ext.client_cert_type_ctos = 0;
+        sc->ext.client_cert_type = TLSEXT_cert_type_x509;
+        return 1;
+    }
+
+    if (!PACKET_as_length_prefixed_1(pkt, &supported_cert_types)) {
+        SSLfatal(sc, SSL_AD_DECODE_ERROR, SSL_R_BAD_EXTENSION);
+        return 0;
+    }
+    if ((len = PACKET_remaining(&supported_cert_types)) == 0) {
+        SSLfatal(sc, SSL_AD_DECODE_ERROR, SSL_R_BAD_EXTENSION);
+        return 0;
+    }
+    if (!PACKET_get_bytes(&supported_cert_types, &data, len)) {
+        SSLfatal(sc, SSL_AD_DECODE_ERROR, SSL_R_BAD_EXTENSION);
+        return 0;
+    }
+    /* client_cert_type: client (peer) has priority */
+    sc->ext.client_cert_type_ctos = reconcile_cert_type(data, len,
+                                                        sc->client_cert_type, sc->client_cert_type_len,
+                                                        &sc->ext.client_cert_type);
+
+    if (sc->ext.client_cert_type_ctos)
+        return 1;
+
+    /* Did not receive an acceptable cert type */
+    SSLfatal(sc, SSL_AD_UNSUPPORTED_CERTIFICATE, SSL_R_BAD_EXTENSION);
```

Comment:
The handshake should perhaps be able to continue unless the server's verification mode is `SSL_VERIFY_PEER | SSL_VERIFY_FAIL_IF_NO_PEER_CERT`. Though, on the other hand, a client that expects to be able to present a cert or RPK to gain the right access level once connected may not be prepared to deal with connections that did not request the credential, but that could happen for reasons unrelated to cert type negotiation. So failing soft (continuing without client certs) may be a good idea.

Reply:
Per RFC7250:
```
   2.  The server supports the extension defined in this document, but
       it does not have any certificate type in common with the client.
       Then, the server terminates the session with a fatal alert of
       type "unsupported_certificate".

Reply:
The RFC text makes sense for the server certificate type, but is not obviously a wise choice for the client certificate type, when the server is willing/able to proceed without requesting a client certificate. The specification was not carefully considered.

Reply:
But it's what the RFC says. The RFC does say to not sent the client type extension if it client auth won't be done, but it does not caveat this scenario with that caveat. So, I'd rather not violate the RFC because we disagree.

Reply:
It is Matt's turn to review. My approval stands, despite any instances of this sort of minor style differences [yes, I would ignore the RFC where the RFC is unnecessarily strict].

Reply:
I'll throw in a check to see if we're doing client auth, if not, then we'll ignore client_cert_type.

Reply:
Perhaps instead of looking at whether the client's PHA extension has been received (it may be after the client cert type extension in the client hello), look at whether the server verification mode includes `SSL_VERIFY_POST_HANDSHAKE` (which is currently documented as optional, but is a better hint that the application will want client certs than the client's abstract willingness to provide them).  The docs can be updated to say that server applications that want to be able to request client RPKs post handshake, must set the flag. Otherwise, they'll only be able to request X.509 certs.

Reply:
I wrote the PHA code, and should've remembered that!
I'm making a "better" check, to basically ignore the error until we know if we're going to be doing client auth, which means, as you point out, after all the extensions are processed.